Reorganize repo

Clonkk opened the pull request:
* Reorganize repo to separate bindings from higher level nim specific API
* Add check for ``len > 0`` in ``send`` before ``copyMem`` 
* Add `[]` and ``events()`` and ``len()`` for poller to simplify polling code (having to manually deal with bitops is poor IMO)
* Add destructor for TConnection
  * Add ownctx field to TConnection to know wether or not the connexion own the context 
  * destructor added behind a ``when defined(gcDestructors)``, close ``PSocket`` and owned ``PContext`` 
  * Add ``sockaddr`` that keep in memory the address of the TConnection
    * Useful because ``disconnect``, ``unbind`` proc somehow needs the address 

The only API change is adding 2 fields to the type ``TConnection``. No proc changed signature - just more proc added to improve the quality of life.

@Araq (or any maintainer that read this) at your leisure, I'd like have your opinion on the change before I spend any more time on this.

Araq commented on zmq/connections.nim:
```diff
@@ -0,0 +1,199 @@
+when defined(gcDestructors):
+  proc close*(c: TConnection)
+  proc `=destroy`(x: var TConnection) =
+    echo "=destroy"
```
debug leftover.

Araq commented on zmq/connections.nim:
```diff
@@ -0,0 +1,199 @@
+    echo "=destroy"
+    # Set linger to 0 to properly drop message
+    setsockopt(x, LINGER, 0.cint)
+    close(x)
```
Instead of closing in the destructor which then cannot really report errors back, assert in the destructor that `close` was called.

Clonkk replied:
I raise an exception if destructor is called without previous manual close. Was that what you hadi n mind ?

Araq replied:
That is what I had in mind.

Araq commented on zmq/poller.nim:
```diff
@@ -0,0 +1,47 @@
+  Poller* = object
+    items*: seq[TPollItem]
+
+proc `[]`*(poller: Poller, idx : int): TPollItem =
```
```suggestion
proc `[]`*(poller: Poller, idx: int): lent TPollItem =
```

Araq commented on zmq/poller.nim:
```diff
@@ -0,0 +1,47 @@
+  Poller* = object
+    items*: seq[TPollItem]
+
+proc `[]`*(poller: Poller, idx : int): lent TPollItem =
```
```suggestion
proc `[]`*(poller: Poller, idx: int): lent TPollItem =
```

Araq commented:
I would go even further in the reorganization:

- Remove the outdated T/P type prefixes.
- Add CI integration.
- Bump version to 1.0.

Clonkk commented:
Which CI is available (Azure or Travis I assume) ? Is there a yaml template it ?

Clonkk commented:
@Araq I think this PR is ready for review.

Since we went for a 1.0 : 
* I renamed obsolete type to ``ZTypeName`` (prefix to avoid name conflict)
* Renamed field with 1 letter name to a more explicit ``context``, ``socket``
* Added rudimentary tests (no auth features tested)

I used github actions based on https://github.com/iffy/install-nim. 
If that's not the preferred way for CI, I'd be happy to change it to whatever is standard for "official" repository.
The github actions simply uses testament to run the tests and examples. 

There is no documentation generation for now as I'm not sure how you want to handle it for this repo  (I usually use github pages on ``docs/`` folder and commit the generated files).

Araq commented on zmq/connections.nim:
```diff
@@ -0,0 +1,257 @@
+  Types definition
+]#
+type
+  EZmq* = object of IOError ## exception that is raised if something fails
```
This should be `ZmqError`

Clonkk replied:
Fixed
